tests/integration/taint_test.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ints
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pulumi/pulumi/pkg/v3/testing/integration"
+)
+
+// TestTaintReplaceResource tests the taint/untaint workflow
+func TestTaintReplaceResource(t *testing.T) {
+	t.Parallel()
+
+	// Create a simple Node.js program that creates a resource
+	dir := t.TempDir()
+
+	// Write a simple Pulumi program
+	program := `
+const pulumi = require("@pulumi/pulumi");
+
+class MyResource extends pulumi.CustomResource {
+    constructor(name, args, opts) {
+        super("test:index:MyResource", name, args, opts);
+    }
+}
+
+const resource1 = new MyResource("myresource", {
+    prop1: "value1",
+});
+
+exports.resource1Id = resource1.id;
+`
+
+	err := os.WriteFile(filepath.Join(dir, "index.js"), []byte(program), 0600)
+	require.NoError(t, err)
+
+	// Write package.json
+	packageJSON := `{
+  "name": "taint-test",
+  "main": "index.js",
+  "dependencies": {
+    "@pulumi/pulumi": "^3.0.0"
+  }
+}
+`
+	err = os.WriteFile(filepath.Join(dir, "package.json"), []byte(packageJSON), 0600)
+	require.NoError(t, err)
+
+	// Write Pulumi.yaml
+	pulumiYAML := `name: taint-test
+runtime: nodejs
+description: Test taint functionality
+`
+	err = os.WriteFile(filepath.Join(dir, "Pulumi.yaml"), []byte(pulumiYAML), 0600)
+	require.NoError(t, err)
+
+	e := &integration.ProgramTestOptions{
+		Dir:             dir,
+		Dependencies:    []string{"@pulumi/pulumi"},
+		Quick:           true,
+		SkipRefresh:     true,
+		NoParallel:      true,
+		PrepareProject:  func(*engine.Projinfo) error { return nil },
+
+		// Run the workflow
+		ExtraRuntimeValidation: func(t *testing.T, stack integration.RuntimeValidationStackInfo) {
+			// Step 1: Get the URN of the resource
+			var resourceURN string
+			for _, res := range stack.Deployment.Resources {
+				if res.Type == "test:index:MyResource" {
+					resourceURN = string(res.URN)
+					break
+				}
+			}
+			require.NotEmpty(t, resourceURN, "MyResource URN not found")
+
+			// Step 2: Taint the resource
+			stdout, stderr, err := stack.RunCommand("pulumi", "state", "taint", resourceURN, "--yes")
```

Comment:
I don't think `stack.RunCommand` exists.  We should probably just use `ptesting.NewEnvironment` for these tests.